Importing required libraries

In [1]:
import numpy as np
import pandas as pd
import cv2
import matplotlib.pyplot as plt

In [2]:
from sklearn.datasets import load_iris

Read the of the data into a dataframe

In [3]:
iris = load_iris()

In [4]:
iris_pd = pd.DataFrame(iris.data,columns=iris.feature_names)
iris_pd

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]

In [5]:
iris.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [6]:
iris_pd['targets'] = iris.target

In [7]:
iris_pd

sepal length (cm)  sepal width (cm)  ...  petal width (cm)  targets
0                  5.1               3.5  ...               0.2        0
1                  4.9               3.0  ...               0.2        0
2                  4.7               3.2  ...               0.2        0
3                  4.6               3.1  ...               0.2        0
4                  5.0               3.6  ...               0.2        0
..                 ...               ...  ...               ...      ...
145                6.7               3.0  ...               2.3        2
146                6.3               2.5  ...               1.9        2
147                6.5               3.0  ...               2.0        2
148                6.2               3.4  ...               2.3        2
149                5.9               3.0  ...               1.8        2

[150 rows x 5 columns]

In [8]:
iris_pd = iris_pd.sample(frac=1)
iris_pd

sepal length (cm)  sepal width (cm)  ...  petal width (cm)  targets
147                6.5               3.0  ...               2.0        2
104                6.5               3.0  ...               2.2        2
146                6.3               2.5  ...               1.9        2
100                6.3               3.3  ...               2.5        2
68                 6.2               2.2  ...               1.5        1
..                 ...               ...  ...               ...      ...
131                7.9               3.8  ...               2.0        2
77                 6.7               3.0  ...               1.7        1
19                 5.1               3.8  ...               0.3        0
30                 4.8               3.1  ...               0.2        0
12                 4.8               3.0  ...               0.1        0

[150 rows x 5 columns]

Making of feature matrix and target matrix

In [9]:
X = iris_pd.loc[:, iris_pd.columns != 'targets'].to_numpy()
y = iris_pd['targets'].to_numpy()

In [10]:
y

array([2, 2, 2, 2, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 2, 1, 0, 0, 0,
       2, 1, 0, 1, 1, 1, 1, 0, 2, 2, 1, 1, 1, 1, 2, 2, 0, 2, 2, 1, 2, 1,
       0, 2, 1, 1, 1, 0, 2, 1, 0, 2, 1, 0, 2, 2, 0, 0, 2, 0, 2, 0, 2, 2,
       0, 0, 2, 2, 2, 2, 2, 1, 0, 1, 2, 2, 0, 1, 2, 1, 0, 0, 2, 1, 2, 0,
       0, 1, 2, 1, 2, 0, 2, 2, 0, 0, 1, 2, 0, 1, 1, 1, 2, 2, 1, 2, 2, 2,
       0, 1, 2, 0, 0, 0, 1, 2, 0, 1, 2, 0, 1, 1, 2, 0, 0, 2, 0, 0, 1, 0,
       1, 1, 2, 1, 0, 1, 0, 0, 1, 1, 1, 0, 2, 2, 1, 0, 0, 0])

To make all features in same range we import StandardScalar of sklearn to do this.



In [11]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X = sc.fit_transform(X)

Importing SVM, Random forest, Logistic Regression classifiers for model selection



In [12]:
from sklearn import svm
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RandomizedSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

Running Grid search for finding optimal model for prediction

In [13]:
grid = {
    'svm':{
        'model': svm.SVC(gamma = 'auto'),
        'par':{
            'C':[1,10,100],
            'kernel':['rbf','linear','poly']
        }
    },
    'RandomForest':{
        'model': RandomForestClassifier(),
        'par':{
            'n_estimators':[20,40,60,80,100]
        }
    },
    'Logistic':{
        'model': LogisticRegression(),
        'par':{
            'C':[5,10,25,50,75,100]
        }    
    }
}

In [14]:
scores = []
for model_name,mp in grid.items():
  model = GridSearchCV(mp['model'],mp['par'],cv= 5)
  model.fit(X,y)
  scores.append({
      'model':model_name,
      'best_score': model.best_score_,
      'best_param': model.best_params_,
  })

In [15]:
k = pd.DataFrame(scores,columns=['model','best_score','best_param'])
k

model  best_score                 best_param
0           svm    0.973333  {'C': 1, 'kernel': 'rbf'}
1  RandomForest    0.940000       {'n_estimators': 20}
2      Logistic    0.966667                   {'C': 5}

As we can see we get highest accuracy when model is SVM with accuracy of 0.973333



In [ ]:
model = svm.SVC(C=1,kernel='rbf')